src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/format-hex/Format-Hex.cs

```diff
@@ -312,6 +261,94 @@ private void ProcessObjectContent(PSObject inputObject)
                     WriteHexidecimal(inputBytes, null, 0);
                 }
             }
+            else
+            {
+                string errorMessage = StringUtil.Format(UtilityCommonStrings.FormatHexTypeNotSupported, obj.GetType());
+                ErrorRecord errorRecord = new ErrorRecord(new ArgumentException(errorMessage),
+                                                            "FormatHexTypeNotSupported",
+                                                            ErrorCategory.InvalidArgument,
+                                                            obj.GetType());
+                WriteError(errorRecord);
+            }
+        }
+
+        private byte[] ConvertToByteArray(object inputObject)
+        {
+            if (inputObject is string str)
+            {
+                return Encoding.GetBytes(str);
+            }
+
+            var baseType = inputObject.GetType();
+            byte[] result = null;
+            int elements = 1;
+            bool isArray = false;
+            bool isBool = false;
+            bool isEnum = false;
+            if (baseType.IsArray)
+            {
+                baseType = baseType.GetElementType();
+                dynamic o = inputObject;
+                elements = (int)o.Length;
+                isArray = true;
+            }
+
+            if (baseType.IsEnum)
+            {
+                baseType = baseType.GetEnumUnderlyingType();
+                isEnum = true;
+            }
+
+            if (baseType.IsPrimitive && elements > 0)
+            {
+                if (baseType == typeof(bool))
+                {
+                    isBool = true;
+                }
+
+                var elementSize = Marshal.SizeOf(baseType);
+                result = new byte[elementSize * elements];
+                if (!isArray)
+                {
+                    inputObject = new object[1] { inputObject };
+                }
+
+                int index = 0;
+                foreach (dynamic obj in (Array)inputObject)
+                {
+                    if (elementSize == 1)
+                    {
+                        result[index] = (byte)obj;
+                    }
+                    else
+                    {
+                        dynamic toBytes;
+                        if (isEnum)
+                        {
+                            toBytes = Convert.ChangeType(obj, baseType);
+                        }
+                        else if (isBool)
+                        {
+                            // bool is 4 bytes apparently
```

Comment:
`bool` is 1 byte, not 4.

There is a difference between `sizeof` and `Marshal.SizeOf`, see  See the [sizeof](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/sizeof) documentation, also this [blog](https://ericlippert.com/2013/06/13/whats-the-difference-sizeof-and-marshal-sizeof/).

[BitConverter.GetBytes(bool)](https://docs.microsoft.com/en-us/dotnet/api/system.bitconverter.getbytes?view=netframework-4.7.2#System_BitConverter_GetBytes_System_Boolean_) works on managed memory and returns a 1 byte array.

I do think this code should use the managed representation for primitive types.

If `Format-Hex` is enhanced to support an arbitrary `ValueType`, then we should probably have a discussion about which representation is more useful because in some cases, the unmanaged layout might be more useful.

Reply:
We could add `-Unmanaged` switch and by default do `managed` output.

Reply:
Possibly, but we could also check the attributes on the struct, e.g. most structs used for interop specify `[StructLayout(LayoutKind.Sequential)]`.

Reply:
It looks unnecessarily complicated, especially since CoreCLR/CoreFX uses some tricks to align structures. Seems this attribute does not always mean interop.

Reply:
So we agree that this should be 1 byte by default?